File: src/domain/dedupe.ts
import { normalizeText } from './normalize'
import type { Halalas } from './money'
import type { IsoDate, MatchingState } from './entities/types'

/**
 * منع التكرار — الدرجات الخمس في spec/05.
 *
 *   ١. بصمة ملف مطابق سبق استيراده  ← تُعالَج في حالة الاستخدام قبل التحليل
 *   ٢. مرجع + هوية حساب متطابقان وتفاصيل متطابقة  ← مكرر مؤكد
 *   ٣. مرجع مطابق وتفاصيل مختلفة                  ← تعارض، لا استبدال صامت
 *   ٤. تشابه بلا مرجع موثوق                       ← مرشح مراجعة، لا حذف تلقائي
 *   ٥. مصدران لعملية واحدة (رسالة + كشف)          ← SourceRecords متعددة
 *
 * ملاحظة حاسمة على الدرجة ٤: «عمليتا مطعم بالمبلغ نفسه قد تكونان حقيقتين» —
 * لذلك التشابه **لا يحذف أبدًا**، يعرض للقرار فقط.
 *
 * بلا مرجع، **سطر الكشف نفسه** يُعرف بالحساب + اليوم + المبلغ + الاتجاه + **الرصيد
 * المعلن بعد الحركة**: عمليتان حقيقيتان متتاليتان لا يتساوى رصيدهما بعدهما. ده بيمسك
 * إعادة استيراد نفس الكشف لما اسم التاجر يتقرا مختلف (HANDOVER §29)، ومن غير رصيد
 * معلن لا يُحكم بالتكرار من اليوم والمبلغ وحدهما.
 */

/** بصمة سطر واردة من الاستيراد. */
export interface DedupeCandidate {
  /** نطاق تفرّد المرجع: مصدر + حساب. مرجع من حسابين مختلفين ليس تكرارًا. */
  smsSource?: boolean
  accountIdentity: string
  sourceReference: string | null
  date: IsoDate
  amountMinor: Halalas
  direction: 'in' | 'out'
  merchantName: string
  rowIndex: number
  /** الرصيد المعلن بعد الحركة في الكشف، إن وُجد. */
  statedBalanceMinor?: Halalas
}

/** سجل موجود مسبقًا للمقارنة. */
export interface ExistingRecord extends DedupeCandidate {
  transactionId: string
}

export interface DedupeVerdict {
  state: MatchingState
  /** سبب القرار بلغة المستخدم. لا حالة بلا تفسير. */
  reason: string
  /** العملية الموجودة المرتبطة، عند التكرار أو التعارض أو التشابه. */
  matchedTransactionId?: string
  /** تفاصيل الاختلاف عند التعارض. */
  conflictFields?: string[]
}

/**
 * مفتاح التفاصيل: تاريخ + مبلغ + اتجاه + اسم تاجر مُطبعَن.
 * يستخدم للتمييز بين «مكرر مؤكد» و«تعارض» عند تطابق المرجع.
 */
export function detailKey(c: DedupeCandidate): string {
  return [c.date, String(c.amountMinor), c.direction, normalizeText(c.merchantName)].join('|')
}

/** مفتاح المرجع مقيّد بهوية الحساب — spec/03: «المرجع الفريد مقيد بالمصدر والحساب». */
export function referenceKey(c: DedupeCandidate): string | null {
  if (!c.sourceReference || !c.sourceReference.trim()) return null
  return `${c.accountIdentity}|${c.sourceReference.trim()}`
}

/** مفتاح سطر الكشف بالرصيد المعلن — null لو مفيش رصيد (رسالة، أو مخطط بلا عمود رصيد). */
export function balanceKey(c: DedupeCandidate): string | null {
  if (c.smsSource || c.statedBalanceMinor === undefined) return null
  return [c.accountIdentity, c.date, c.amountMinor, c.direction, c.statedBalanceMinor].join('|')
}

function diffFields(a: DedupeCandidate, b: DedupeCandidate): string[] {
  const out: string[] = []
  if (a.date !== b.date) out.push('التاريخ')
  if (a.amountMinor !== b.amountMinor) out.push('المبلغ')
  if (a.direction !== b.direction) out.push('اتجاه الحركة')
  if (normalizeText(a.merchantName) !== normalizeText(b.merchantName)) out.push('اسم التاجر')
  return out
}

/** فهرس للمقارنة السريعة، يُبنى مرة قبل فحص الدفعة. */
export interface DedupeIndex {
  byAmountDay: Map<string, ExistingRecord[]>
  byReference: Map<string, ExistingRecord>
  byDetail: Map<string, ExistingRecord[]>
  byBalance: Map<string, ExistingRecord>
}

export function buildDedupeIndex(existing: readonly ExistingRecord[]): DedupeIndex {
  const byAmountDay = new Map<string, ExistingRecord[]>()
  const byReference = new Map<string, ExistingRecord>()
  const byDetail = new Map<string, ExistingRecord[]>()
  const byBalance = new Map<string, ExistingRecord>()
  for (const record of existing) {
    const dayKey = amountDayKey(record)
    byAmountDay.set(dayKey, [...(byAmountDay.get(dayKey) ?? []), record])
    const ref = referenceKey(record)
    if (ref) byReference.set(ref, record)
    const balance = balanceKey(record)
    if (balance && !byBalance.has(balance)) byBalance.set(balance, record)
    const key = detailKey(record)
    const list = byDetail.get(key)
    if (list) list.push(record)
    else byDetail.set(key, [record])
  }
  return { byReference, byDetail, byAmountDay, byBalance }
}

/**
 * يصنّف سطرًا واحدًا مقابل ما هو موجود.
 * الدالة نقية ولا تعدّل الفهرس — الإضافة مسؤولية المستدعي بعد القرار.
 */
export function classifyCandidate(
  candidate: DedupeCandidate,
  index: DedupeIndex,
): DedupeVerdict {
  const ref = referenceKey(candidate)

  if (ref) {
    const existing = index.byReference.get(ref)
    if (existing) {
      const differences = diffFields(candidate, existing)
      // ─── الدرجة ٢: مرجع + حساب + تفاصيل متطابقة ───
      if (differences.length === 0) {
        return {
          state: 'duplicate',
          reason: `نفس المرجع «${candidate.sourceReference}» ونفس التفاصيل — العملية دي متسجلة قبل كده`,
          matchedTransactionId: existing.transactionId,
        }
      }
      // ─── الدرجة ٣: مرجع مطابق وتفاصيل مختلفة ← تعارض ───
      return {
        state: 'conflict',
        reason:
          `نفس المرجع «${candidate.sourceReference}» بس ${differences.join(' و')} مختلف. ` +
          `مش هنكتب فوق القديم من غير قرارك.`,
        matchedTransactionId: existing.transactionId,
        conflictFields: differences,
      }
    }
  }

  // ─── نفس سطر الكشف بالرصيد المعلن، حتى لو اسم التاجر اتقرا مختلف ───
  const sameLine = balanceKey(candidate)
  const lineMatch = sameLine ? index.byBalance.get(sameLine) : undefined
  if (lineMatch) {
    return {
      state: 'duplicate',
      reason:
        'نفس سطر الكشف: نفس اليوم والمبلغ والاتجاه والرصيد بعد العملية في نفس الحساب — ' +
        'متسجلة قبل كده حتى لو الاسم مقروء بشكل مختلف',
      matchedTransactionId: lineMatch.transactionId,
    }
  }

  if (ref) {
    const related = smsSimilarity(candidate, index)
    if (related) return related
    // مرجع جديد بالكامل — لا يُفحص التشابه لأن المرجع دليل موثوق
    return { state: 'new', reason: 'مرجع جديد مش موجود قبل كده' }
  }

  const related = smsSimilarity(candidate, index)
  if (related) return related
  // ─── الدرجة ٤: بلا مرجع موثوق — تشابه يحتاج قرارًا، لا حذفًا ───
  const similar = index.byDetail.get(detailKey(candidate))
  if (similar && similar.length > 0) {
    return {
      state: 'similar',
      reason:
        'نفس التاجر والمبلغ والتاريخ والاتجاه، ومفيش مرجع بنكي يأكد. ' +
        'ممكن تكون نفس العملية وممكن تكون عملية تانية حقيقية — محتاجة قرارك.',
      matchedTransactionId: similar[0].transactionId,
    }
  }

  return { state: 'new', reason: 'عملية جديدة، مفيش ما يشبهها' }
}

/**
 * بصمة الملف — الدرجة ١. تجزئة FNV-1a 64-bit بلا مكتبة.
 * الغرض تمييز «نفس الملف بالضبط»، لا الحماية من العبث.
 */
export function hashContent(content: string): string {
  let hi = 0xcbf29ce4 >>> 0
  let lo = 0x84222325 >>> 0
  for (let i = 0; i < content.length; i++) {
    const code = content.charCodeAt(i)
    lo ^= code & 0xff
    hi ^= (code >>> 8) & 0xff
    // الضرب في 16777619 بأجزاء 16-بت لتفادي فقد الدقة
    const l = Math.imul(lo, 16777619) >>> 0
    const h = (Math.imul(hi, 16777619) + Math.imul(lo, 1099511628211 % 4294967296)) >>> 0
    lo = l
    hi = h
  }
  return (hi.toString(16).padStart(8, '0') + lo.toString(16).padStart(8, '0')).toUpperCase()
}

function amountDayKey(c:DedupeCandidate):string {
 return [c.accountIdentity,c.date,c.amountMinor,c.direction].join('|')
}
function smsSimilarity(c:DedupeCandidate,index:DedupeIndex):DedupeVerdict|null {
 const matches=index.byAmountDay.get(amountDayKey(c))??[]
 const match=matches.find(r=>c.smsSource||r.smsSource)
 if(!match)return null
 return {state:'similar',matchedTransactionId:match.transactionId,reason:'فيه عملية بنفس اليوم والمبلغ والاتجاه في نفس المحفظة، وأحد المصدرين رسالة بنك. قد تكون نفس العملية باسم مختلف؛ غير مختارة للإضافة حتى تراجعها.'}
}

export function importFingerprint(content:string,accountIdentity:string):string {
 return hashContent(JSON.stringify([accountIdentity,content]))
}

File: native-app/core/src/commonMain/kotlin/app/masroufy/core/Dedupe.kt
package app.masroufy.core

/**
 * منع التكرار — نقل `src/domain/dedupe.ts` (الدرجات الخمس في spec/05).
 * التشابه **ما بيمسحش أبدًا**، بيتعرض للقرار بس. ومن غير مرجع، سطر الكشف بيتعرف بالحساب + اليوم + المبلغ
 * + الاتجاه + **الرصيد المعلن بعد الحركة** (مطابقة واحد لواحد).
 */
enum class MatchingState(val wire: String) {
    NEW("new"), DUPLICATE("duplicate"), SIMILAR("similar"), CONFLICT("conflict"), INVALID("invalid");

    companion object {
        fun fromWire(wire: String): MatchingState = entries.first { it.wire == wire }
    }
}

data class DedupeCandidate(
    val accountIdentity: String,
    val sourceReference: String?,
    val date: IsoDate,
    val amountMinor: Halalas,
    val direction: Direction,
    val merchantName: String,
    val rowIndex: Int,
    /** الرصيد المعلن بعد الحركة في الكشف، لو موجود. */
    val statedBalanceMinor: Halalas? = null,
    /** المصدر رسالة بنك (مفيش رصيد سطر يتطابق عليه). */
    val smsSource: Boolean = false,
)

data class ExistingRecord(val candidate: DedupeCandidate, val transactionId: String)

data class DedupeVerdict(
    val state: MatchingState,
    /** السبب بلغة المستخدم — مفيش حالة من غير تفسير. */
    val reason: String,
    val matchedTransactionId: String? = null,
    val conflictFields: List<String>? = null,
    /** مفتاح الرصيد اللي اتطابق — المستدعي بيزوّد عدّاده (كل سجل موجود بيبلع صف وارد واحد بس). */
    val matchedBalanceKey: String? = null,
)

/** تاريخ + مبلغ + اتجاه + اسم تاجر مطبعن — بيفرّق «مكرر» عن «تعارض» عند تطابق المرجع. */
fun detailKey(c: DedupeCandidate): String = listOf(c.date, c.amountMinor.toString(), c.direction.wire, normalizeText(c.merchantName)).joinToString("|")

/** المرجع مقيد بهوية الحساب (spec/03). */
fun referenceKey(c: DedupeCandidate): String? {
    val ref = c.sourceReference ?: return null
    val trimmed = JsText.trim(ref)
    return if (trimmed.isEmpty()) null else "${c.accountIdentity}|$trimmed"
}

/** مفتاح سطر الكشف بالرصيد المعلن — null لو مفيش رصيد (رسالة، أو ملف من غير عمود رصيد). */
fun balanceKey(c: DedupeCandidate): String? {
    if (c.smsSource || c.statedBalanceMinor == null) return null
    return listOf(c.accountIdentity, c.date, c.amountMinor, c.direction.wire, c.statedBalanceMinor).joinToString("|")
}

private fun amountDayKey(c: DedupeCandidate) = listOf(c.accountIdentity, c.date, c.amountMinor, c.direction.wire).joinToString("|")

private fun diffFields(a: DedupeCandidate, b: DedupeCandidate): List<String> = buildList {
    if (a.date != b.date) add(uiText(TextKey.DIFF_DATE))
    if (a.amountMinor != b.amountMinor) add(uiText(TextKey.DIFF_AMOUNT))
    if (a.direction != b.direction) add(uiText(TextKey.DIFF_DIRECTION))
    if (normalizeText(a.merchantName) != normalizeText(b.merchantName)) add(uiText(TextKey.DIFF_MERCHANT))
}

class DedupeIndex internal constructor(
    internal val byAmountDay: Map<String, List<ExistingRecord>>,
    internal val byReference: Map<String, ExistingRecord>,
    internal val byDetail: Map<String, List<ExistingRecord>>,
    /** **قايمة** لكل مفتاح: الراجحي بيطبع رصيد نهاية اليوم على كل سطور اليوم. */
    internal val byBalance: Map<String, List<ExistingRecord>>,
)

fun buildDedupeIndex(existing: List<ExistingRecord>): DedupeIndex {
    val byAmountDay = mutableMapOf<String, MutableList<ExistingRecord>>()
    val byReference = mutableMapOf<String, ExistingRecord>()
    val byDetail = mutableMapOf<String, MutableList<ExistingRecord>>()
    val byBalance = mutableMapOf<String, MutableList<ExistingRecord>>()
    for (record in existing) {
        val c = record.candidate
        byAmountDay.getOrPut(amountDayKey(c)) { mutableListOf() }.add(record)
        referenceKey(c)?.let { byReference[it] = record }
        balanceKey(c)?.let { byBalance.getOrPut(it) { mutableListOf() }.add(record) }
        byDetail.getOrPut(detailKey(c)) { mutableListOf() }.add(record)
    }
    return DedupeIndex(byAmountDay, byReference, byDetail, byBalance)
}

private fun smsSimilarity(c: DedupeCandidate, index: DedupeIndex): DedupeVerdict? {
    val match = index.byAmountDay[amountDayKey(c)].orEmpty().firstOrNull { c.smsSource || it.candidate.smsSource } ?: return null
    return DedupeVerdict(
        MatchingState.SIMILAR,
        uiText(TextKey.DEDUPE_SMS_SIMILAR),
        matchedTransactionId = match.transactionId,
    )
}

/** بيصنّف سطر واحد قصاد الموجود. نقية — ما بتعدّلش الفهرس. `consumedBalance` قراية بس. */
fun classifyCandidate(candidate: DedupeCandidate, index: DedupeIndex, consumedBalance: Map<String, Int>? = null): DedupeVerdict {
    val ref = referenceKey(candidate)
    if (ref != null) {
        val existing = index.byReference[ref]
        if (existing != null) {
            val differences = diffFields(candidate, existing.candidate)
            if (differences.isEmpty()) {
                return DedupeVerdict(
                    MatchingState.DUPLICATE,
                    uiText(TextKey.DEDUPE_SAME_REFERENCE, candidate.sourceReference ?: ""),
                    matchedTransactionId = existing.transactionId,
                )
            }
            return DedupeVerdict(
                MatchingState.CONFLICT,
                uiText(TextKey.DEDUPE_REFERENCE_CONFLICT, candidate.sourceReference ?: "", differences.joinToString(uiText(TextKey.JOIN_AND))),
                matchedTransactionId = existing.transactionId,
                conflictFields = differences,
            )
        }
    }

    val sameLine = balanceKey(candidate)
    if (sameLine != null) {
        val alreadyUsed = consumedBalance?.get(sameLine) ?: 0
        val lineMatch = index.byBalance[sameLine].orEmpty().getOrNull(alreadyUsed)
        if (lineMatch != null) {
            return DedupeVerdict(
                MatchingState.DUPLICATE,
                uiText(TextKey.DEDUPE_SAME_STATEMENT_LINE),
                matchedTransactionId = lineMatch.transactionId,
                matchedBalanceKey = sameLine,
            )
        }
    }

    smsSimilarity(candidate, index)?.let { return it }
    if (ref != null) return DedupeVerdict(MatchingState.NEW, uiText(TextKey.DEDUPE_NEW_REFERENCE))

    val similar = index.byDetail[detailKey(candidate)]
    if (!similar.isNullOrEmpty()) {
        return DedupeVerdict(
            MatchingState.SIMILAR,
            uiText(TextKey.DEDUPE_SIMILAR_NO_REFERENCE),
            matchedTransactionId = similar[0].transactionId,
        )
    }
    return DedupeVerdict(MatchingState.NEW, uiText(TextKey.DEDUPE_NEW))
}

/**
 * بصمة الملف (الدرجة ١) — نفس FNV-1a بتاع التطبيق الحالي بالظبط، على وحدات UTF-16.
 * ضرب الـInt في كوتلن بيلف زي `Math.imul`، و`1099511628211 % 2^32` = 435.
 */
fun hashContent(content: String): String {
    var hi = 0xcbf29ce4.toInt()
    var lo = 0x84222325.toInt()
    for (ch in content) {
        val code = ch.code
        lo = lo xor (code and 0xff)
        hi = hi xor ((code ushr 8) and 0xff)
        val l = lo * 16777619
        val h = hi * 16777619 + lo * 435
        lo = l
        hi = h
    }
    return (hi.toUInt().toString(16).padStart(8, '0') + lo.toUInt().toString(16).padStart(8, '0')).uppercase()
}

fun importFingerprint(content: String, accountIdentity: String): String =
    hashContent("[${JsText.jsonString(accountIdentity)},${JsText.jsonString(content)}]")
